drt: avoid creating unnecessary patch metal

eder-matheus opened the pull request:
Fixes OR-1638

maliberty commented on src/drt/src/dr/FlexDR_maze.cpp:
```diff
@@ -2849,7 +2849,7 @@ void FlexDRWorker::routeNet_prepAreaMap(drNet* net,
       auto it = areaMap.find(mIdx);
       if (it != areaMap.end()) {
-        it->second = max(it->second, ap->getBeginArea());
+        areaMap[mIdx] = max(it->second, ap->getBeginArea());
```
why the change?

maliberty commented on src/drt/src/dr/FlexDR_maze.cpp:
```diff
@@ -3088,44 +3105,53 @@ void FlexDRWorker::routeNet_postAstarPatchMinAreaVio(
-        currArea = getHalfViaEncArea(
-            prevIdx.z() - 1, true, net->getFrNet()->getNondefaultRule());
+      if (nextIdx.z() < currIdx.z()) {
+        // get the bottom layer box of the current via to initialize the area
+        // for the next shape
+        currArea
+            = getHalfViaEncArea(
+                  currIdx.z() - 1, true, net->getFrNet()->getNondefaultRule())
+              * 2;
```
Is this correct if the next shape is a wire?

eder-matheus replied:
I've tried to compensate for it in this line:
https://github.com/The-OpenROAD-Project/OpenROAD/pull/2581/files/b9d5dfea7a109560dcdc3f6625afec2c338280ee#diff-3b0ee5af511f4838323555d94ad93cc7da9df71a64bf049fa6b634da77978e64R3141

maliberty replied:
What happens when there are multiple wire segments in series.  It seems "currArea /= 2;" would reduce the area of the previous segment not only the via.

eder-matheus replied:
I've updated the code to perform `currArea /= 2;` only when the last object is not a wire. I believe this will prevent reducing the area of all the previous segments.

maliberty replied:
My guess is this won't hurt anything.  Do you want to roll the dice or do another secure CI?

eder-matheus replied:
We've waited a long time for this PR, and waiting a few more hours is fine. I'll start the secure-ci now.

eder-matheus commented:
@maliberty Secure-ci is failing for ca53, with the same 2 DRCs we have on master. It also fails for gf12/swerv_wrapper, but on a CTS metric that is not related to my change. I think it's safe to merge this PR.

maliberty commented on src/drt/src/dr/FlexDR_maze.cpp:
```diff
@@ -2982,28 +2982,45 @@ void FlexDRWorker::routeNet_postAstarPatchMinAreaVio(
   }
   frCoord startViaHalfEncArea = 0, endViaHalfEncArea = 0;
-  FlexMazeIdx prevIdx = points[0], currIdx;
+  FlexMazeIdx currIdx = points[0], nextIdx;
   int i;
   int prev_i = 0;  // path start point
+  bool prev_is_wire = true;
```
The terminology is getting mixed.  You changed prev/curr to curr/next but kept prev_i/prev_is_wire.

eder-matheus replied:
I've kept prev_i and prev_is_wire because they're actually referring to the previous objects. If I understand the code correctly, the vars that I've changed to curr/next didn't have a good naming with prev/curr.

maliberty commented on src/drt/src/dr/FlexDR_maze.cpp:
```diff
@@ -2982,28 +2982,45 @@ void FlexDRWorker::routeNet_postAstarPatchMinAreaVio(
-        currArea += getHalfViaEncArea(
-            prevIdx.z() - 1, false, net->getFrNet()->getNondefaultRule());
+      // add curr via enclosure
+      if (nextIdx.z() < currIdx.z()) {
```
The if and else blocks are nearly identical.  Rather than having two copies, just assign two variables and use a common block.  "currIdx.z() - 1, false" vs "currIdx.z(), true" appears to be the only difference.

maliberty commented:
Please avoid a force push after a review.  You can see changes since the last review afterwards.

eder-matheus commented:
> Please avoid a force push after a review. You can see changes since the last review afterwards.

Didn't know about it. My bad! I'll avoid rebasing and force-pushing in the future.

maliberty commented:
Prior to review that's fine.

On Tue, Jan 10, 2023 at 1:55 PM Eder Monteiro ***@***.***>
wrote:

> Please avoid a force push after a review. You can see changes since the
> last review afterwards.
>
> Didn't know about it. My bad! I'll avoid rebasing and force-pushing in the
> future.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/The-OpenROAD-Project/OpenROAD/pull/2581#issuecomment-1377942938>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFZ5KWMQML3U5TBMGKPC2LWRXLGRANCNFSM6AAAAAASV7NA6I>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
